GitUI/UserControls/RevisionGrid/RevisionGridControl.cs

```diff
@@ -1537,6 +1506,82 @@ static IEnumerable<ObjectId> TryGetParents(GitModule module, FilterInfo filterIn
             }
         }
 
+        /// <summary>
+        /// Returns the name of a file in a specific revision (following renames and merge commits)
+        /// </summary>
+        public string? GetRevisionFileName(string path, ObjectId? objectId)
+        {
+            if (FilePathByObjectId is null || objectId is null)
+            {
+                return null;
+            }
+
+            if (FilePathByObjectId.TryGetValue(objectId, out string? fileName))
+            {
+                return fileName;
+            }
+
+            const string startOfObjectId = "????";
+
+            GitArgumentBuilder args = new("log")
+            {
+                // --name-only will list each filename on a separate line, ending with an empty line
+                // Find start of a new commit with a sequence impossible in a filename
+                $"--format=\"{startOfObjectId}%H\"",
+                "--name-only",
+                "--follow",
+                "--diff-merges=separate",
+                FindRenamesAndCopiesOpts(),
+                $"{objectId}^1..{objectId}",
+                "--",
+                path,
+            };
+
+            return ParseFileNames(args, startOfObjectId).FirstOrDefault(i => i.revision == objectId.ToString()).fileName;
+        }
+
+        private IEnumerable<(string revision, string fileName)> ParseFileNames(GitArgumentBuilder args, string startOfObjectId)
+        {
+            ExecutionResult result = Module.GitExecutable.Execute(args, outputEncoding: GitModule.LosslessEncoding, throwOnErrorExit: false);
+            LazyStringSplit lines = result.StandardOutput.LazySplit('\n');
+
+            // TODO Check the exit code and warn the user that rename detection could not be done.
+
+            ObjectId currentObjectId = null;
+
+            foreach (string? line in lines.Select(GitModule.ReEncodeFileNameFromLossless))
+            {
+                if (string.IsNullOrEmpty(line))
+                {
+                    // empty line after sha
+                    continue;
+                }
+
+                if (line.StartsWith(startOfObjectId))
+                {
+                    if (line.Length < ObjectId.Sha1CharCount + startOfObjectId.Length
+                        || !ObjectId.TryParse(line, offset: startOfObjectId.Length, out currentObjectId))
+                    {
+                        // Parse error, ignore
+                        currentObjectId = null;
+                    }
+
+                    continue;
+                }
+
+                if (currentObjectId == null)
+                {
+                    // Parsing has failed, ignore
+                    continue;
+                }
+
+                // Add only the first file to the dictionary
+                FilePathByObjectId?.TryAdd(currentObjectId, line);
+
+                yield return (currentObjectId.ToString(), line);
```

Comment:
Why are we reallocating `currentObjectId` as a string? In the callsite we reallocating another ObjectId instace as a string to make a comparison.